packages/appium/lib/schema/schema.js

```diff
@@ -0,0 +1,571 @@
+// @ts-check
+
+import Ajv from 'ajv';
+import addFormats from 'ajv-formats';
+import _ from 'lodash';
+import path from 'path';
+import {DRIVER_TYPE, PLUGIN_TYPE} from '../extension-config';
+import {ReadonlyMap} from '../utils';
+import appiumConfigSchema from './appium-config-schema';
+import {ArgSpec, SERVER_PROP_NAME, APPIUM_CONFIG_SCHEMA_ID} from './arg-spec';
+import {keywords} from './keywords';
+
+/**
+ * Extensions that an extension schema file can have.
+ */
+export const ALLOWED_SCHEMA_EXTENSIONS = new Set(['.json', '.js', '.cjs']);
+
+/**
+ * A wrapper around Ajv and schema-related functions.
+ *
+ * Should have been named Highlander, because _there can only be one_
+ */
+class AppiumSchema {
+  /**
+   * A mapping of unique argument IDs to their corresponding {@link ArgSpec}s.
+   *
+   * An "argument" is a CLI argument or a config property.
+   *
+   * Used to provide easy lookups of argument metadata when converting between different representations of those arguments.
+   * @private
+   * @type {ReadonlyMap<string,ArgSpec>}
+   */
+  _argSpecs = new ReadonlyMap();
+
+  /**
+   * A map of extension types to extension names to schema objects.
+   *
+   * This data structure is used to ensure there are no naming conflicts. The schemas
+   * are stored here in memory until the instance is _finalized_.
+   * @private
+   * @type {Record<ExtensionType,Map<string,SchemaObject>>}
+   */
+  _registeredSchemas = {[DRIVER_TYPE]: new Map(), [PLUGIN_TYPE]: new Map()};
+
+  /**
+   * Ajv instance
+   *
+   * @private
+   * @type {Ajv}
+   */
+  _ajv;
+
+  /**
+   * Singleton instance.
+   * @private
+   * @type {AppiumSchema}
+   */
+  static _instance;
+
+  /**
+   * Lookup of schema IDs to finalized schemas.
+   *
+   * This does not include references, but rather the root schemas themselves.
+   * @private
+   * @type {Record<string,StrictSchemaObject>?}
+   */
+  _finalizedSchemas = null;
+
+  /**
+   * Initializes Ajv, adds standard formats and our custom keywords.
+   * @see https://npm.im/ajv-formats
+   * @private
+   */
+  constructor () {
+    this._ajv = AppiumSchema._instantiateAjv();
+  }
+
+  /**
+   * Factory function for {@link AppiumSchema} instances.
+   *
+   * Returns a singleton instance if one exists, otherwise creates a new one.
+   * Binds public methods to the instance.
+   * @returns {AppiumSchema}
+   */
+  static create () {
+    const instance = AppiumSchema._instance ?? new AppiumSchema();
+    AppiumSchema._instance = instance;
+
+    _.bindAll(instance, [
```

Comment:
why do we need binding?

Reply:
mainly for ergonomics.  we have a singleton instance of `AppiumSchema`.  we could export the instance, but consumers (any module which imports this one) don't really care about the instance, and just want methods.  These methods of course have references to `this`, which need to be maintained if we call the methods directly.  another way could have just been to create lambdas for each exported function, e.g.:

```js
export const finalizeSchema = (...args) => appiumSchema.finalize(...args);
```

but I thought it'd be more concise to just bind everything.

Reply:
oh, i thought they were all member methods so didn't need binding. guess i read the scoping wrong